Fix for "dune external-lib-deps" command

moyodiallo opened the pull request:
The new version of dune (from 3.0.0) removed "dune external-lib-deps" and this is a patch that work with a patch of dune from "https://github.com/moyodiallo/dune/tree/opam-dune-lint".

There's already a PR(https://github.com/ocaml/dune/pull/6045) in dune about this patch, but not guarantee to be merged. The command may change also.

This could be a temporary solution for our CI projects.

@tmcgilchrist

moyodiallo committed:
Replace "dune external-lib-deps" command

The new version of dune (from 3.0.0) removed "dune external-lib-deps" and
this is a patch that work with a patch of dune from
"https://github.com/moyodiallo/dune/tree/opam-dune-lint".

There's already a PR(https://github.com/ocaml/dune/pull/6045) in dune
about this patch, but not guarantee to be merged. The command may change also.

moyodiallo committed:
Replace dune-lang by sexplib

The library dune-lang is not exposed anymore. It started at 3.0.0 .

moyodiallo committed:
Replacing 'dune external-lib-deps'

In this stage, opam-dune-lint can only works with this dune PR
"https://github.com/ocaml/dune/pull/6839".

Not ready yet, works is needed at this stage.

moyodiallo commented:
At the stage `opam-dune-lint` works using this dune PR https://github.com/ocaml/dune/pull/6839.

This particular example is left for now, because `dune describe external-lib-deps`(far different to `dune external-lib-deps`) does not resolve it.
```dune
(executable{s}
  (name{s} prog)
  (libraries a b))

(install 
  (section bin)
  (package pkg)
  (files prog.exe))
```
- The solution that I have at the moment, is to solve those cases in `opam-dune-lint`.

tmcgilchrist commented on .ocamlformat:
```diff
@@ -0,0 +1,6 @@
+version = 0.24.1
+ocaml-version = 4.12.0
```
```suggestion
ocaml-version = 4.10.0
```
The dune-project file uses this version as the earliest supported version. The ocamlformat should be consistent with that so we don't accidentally reformat code using newer (un-supported) constructs.

tmcgilchrist commented on dune_items.ml:
```diff
@@ -0,0 +1,66 @@
+  type t = Required | Optional
+
+  let merge x y =
+    match (x, y) with Required, _ | _, Required -> Required | _ -> Optional
```
```suggestion
    match (x, y) with Required, _ | _, Required -> Required | Optional, Optional -> Optional
```
Be explicit here with the underscore type.

tmcgilchrist commented on dune_items.ml:
```diff
@@ -0,0 +1,66 @@
+  let dump = { names = []; package = None; external_deps = []; source_dir = "" }
+end
+
+open Sexp
```
Could you be more consistent with the Sexp open and Sexp module prefix?
Sometimes this uses `open Sexp` while other places uses a module prefix. It would be clearer to consistently use one or the other.

tmcgilchrist commented on dune-project:
```diff
@@ -16,7 +15,6 @@
   (sexplib (>= v0.14.0))
   (cmdliner (>= 1.1.0))
-  (dune-private-libs (and (>= 2.8.0) (< 3.0)))
   (dune (< 3.0))
```
This should target a newer version of dune and add the `stdune` library dependency to match.

tmcgilchrist commented:
This could use more tests for the dune-file parsing since we no longer use dune's parser for files. 
There is a risk that newer versions of dune introduce breaking format changes or interpretations of previous fields.

talex5 commented on dune-project:
```diff
@@ -1,6 +1,5 @@
 (lang dune 2.7)
 (name opam-dune-lint)
-(formatting disabled)
```
Why did you remove this? This project isn't using ocamlformat and turning that on has nothing to do with this PR.

moyodiallo replied:
True, is your concern the project is intended to not use ocamlformat or this has to be addressed in another PR ?

talex5 replied:
The project does not intend to use ocamlformat, which is why the line was there.

(If a new maintainer takes over this project, they may decide to change the policy, but we can't have everyone who contributes a fix to a project also changing the rest of the project to their own preferred style.)

moyodiallo replied:
I get it, it was not my intention to change the policy and the style I get it from solver-service https://github.com/ocurrent/solver-service/blob/main/.ocamlformat.

moyodiallo committed:
Remove 'dune describe' command

The command 'dune describe external-lib-deps' is used to print
out external-libs and does not print an internal lib.

moyodiallo committed:
Get the correct libs for target: @test.

moyodiallo committed:
Format before writing 'dune-project' file

moyodiallo committed:
Fix test output format

moyodiallo committed:
Clean some part of the code

moyodiallo commented:
I rewrite the PR, there was some error that I made.

At this stage there's 2 things left:
- [x] correcting the test about tests/test_vendoring.t : It's to consider a public executable could have an internal private lib as a dependency which have an external dep lib.
- [x] the install case https://github.com/ocurrent/opam-dune-lint/pull/46#issuecomment-1381596178
- [x] Take in consideration this case https://github.com/ocaml/dune/issues/3499

moyodiallo committed:
Reorganize the code and improve the command

* The command "dune describe package-entries" is used to know if a private
 executable is going to be installed.

* Resolve all the internal dependencies.

moyodiallo commented:
At this stage,  it could be tested with this branch https://github.com/moyodiallo/dune/tree/opam-dune-lint-testing.
before the merge of https://github.com/ocaml/dune/pull/7480 and https://github.com/ocaml/dune/pull/7478.

moyodiallo committed:
refactor: Avoid a recursive resolve on a public library or executable

moyodiallo committed:
refactor the source code

moyodiallo committed:
When an executable is copied before installed in dune

This case is a hack to make dependency on an executable. It's trying to
copy in another name before to install it. OCaml-CI raised this issue
which can be found here `https://github.com/ocaml/dune/issues/3499`

This commit try solve the case but not perfect because to get the rules,
we are parsing the dune source and there's some limitation on that, like
dune variables.

moyodiallo committed:
Refactor: Add the extensions of executables

Add the extensions of executables and fix the parsing format of dune
commands output.

moyodiallo commented:
At this stage all the concerns above is addressed. The tests are OK with this branch of dune https://github.com/moyodiallo/dune/tree/opam-dune-lint-testing. The branch contains the last update of `dune describe external-lib-deps` and `dune describe package-entries` commands.

MisterDA commented:
From afar I think you should merge the calls to `List.map`, this way you only walk the list once and avoid creation of intermediary lists.
```ocaml
List.map f xs |> List.map g
(* can be rewritten as *)
List.map (fun x -> x |> f |> g) xs
```

Also, `List.concat_map f l` gives the same result as [List.concat](https://v2.ocaml.org/api/List.html#VALconcat) ([List.map](https://v2.ocaml.org/api/List.html#VALmap) f l), and is tail-recursive.

moyodiallo committed:
Refactor the code

moyodiallo committed:
Remove "dune build" and improve parsing copy

Building to upgrade the opam files is replaced by another dune command:
`dune describe opam-files` which give all updated opam-files.

Improve parsing dune file for dune copy rule

moyodiallo committed:
Add tests for "dune describe" command

moyodiallo commented:
`opam-dune-lint` does not use `dune build` to update opam files. It uses `dune describe opam-files` now.

MisterDA commented on deps.ml:
```diff
@@ -0,0 +1,141 @@
+    in
+    let d_items_lib =
+      d_items
+      |> List.filter is_lib_item
+      |> List.map (fun d_item ->
```
You may want to use `List.filter_map`.

MisterDA commented on deps.ml:
```diff
@@ -0,0 +1,141 @@
+        else begin
+          Hashtbl.add acc (get_name item) item;
+          (get_item item).internal_deps
+          |> List.filter (fun (_, k) -> Kind.is_required k)
+          |> List.filter_map (fun (name, _) ->
```
You could merge these two.

MisterDA commented on deps.ml:
```diff
@@ -0,0 +1,141 @@
+
+let lib_deps ~pkg ~target =
+  get_dune_items (Hashtbl.create 10) ~pkg ~target
+  |> List.map Describe_external_lib.get_item
+  |> List.fold_left (fun acc (item:Describe_external_lib.item) ->
+      List.map (fun dep -> (fst dep, item.source_dir)) item.external_deps @ acc) []
```
Not sure, but maybe you want `List.fold_left_map`?

moyodiallo replied:
It is different, no need for `fold_left_map`.

moyodiallo committed:
Refactor the code

MisterDA commented on deps.ml:
```diff
@@ -0,0 +1,144 @@
+
+type t = Dir_set.t Libraries.t
+
+let dune_describe_external_lib_deps = Bos.Cmd.(v "dune" % "describe" % "external-lib-deps")
+
+let dune_describe_entries = Bos.Cmd.(v "dune" % "describe" % "package-entries")
```
I think that these should not be top-level expressions. They are evaluated when the module is loaded, and if they fail, then the user cannot call simple things such as `opam-dune-lint --help`. They should be wrapped in `unit -> x` functions, or made lazy.
```suggestion
let dune_describe_external_lib_deps () = Bos.Cmd.(v "dune" % "describe" % "external-lib-deps")

let dune_describe_entries () = Bos.Cmd.(v "dune" % "describe" % "package-entries")
```

moyodiallo replied:
That is right, I've already plan to change that, I noticed it. Your suggestion for making them lazy is :ok_hand:

MisterDA commented on dune_project.ml:
```diff
@@ -1,76 +1,74 @@
 open Types
 
-type t = Dune_lang.t list
+module Deps = struct
+  include Deps
+end
```
This is not necessary and could be removed by tweaking a bit the module definitions. Or with this, if I'm not mistaken.
```suggestion
module Deps = Deps
```

MisterDA reviewed (changes requested):
I think you should rename `dune_project.{ml,mli}` to something else, like `opam_dune_lint.{ml,mli}`.

moyodiallo commented:
> I think you should rename `dune_project.{ml,mli}` to something else, like `opam_dune_lint.{ml,mli}`.

Why something else, the purpose of the module is to upgrade `dune-project` file.

MisterDA commented:
> > I think you should rename `dune_project.{ml,mli}` to something else, like `opam_dune_lint.{ml,mli}`.
> 
> Why something else, the purpose of the module is to upgrade `dune-project` file.

oh, my bad! thanks for the explanation.

moyodiallo committed:
Make the top level expressions lazy and that also increase performance

moyodiallo committed:
Add copy rules when 'dune' file exist

moyodiallo committed:
Add a correction about unresolved entries

The entries are from the command "dune describe package-entries".

moyodiallo committed:
The description is no longer valid

moyodiallo commented:
Ready to go after the merge of
- [x] https://github.com/ocaml/dune/pull/7480
- [x] https://github.com/ocaml/dune/pull/7478

_Another review or testing is welcome._

moyodiallo committed:
Upgrade to match the dune commands that used

moyodiallo committed:
Update dependencies

tmcgilchrist commented on opam-dune-lint.opam:
```diff
@@ -9,12 +9,12 @@ license: "ISC"
-  "dune" {>= "2.7" & < "3.0"}
-  "ocaml" {>= "4.10.0"}
+  "stdune" {>= "3.10.0"}
+  "ocaml" {>= "4.13.0"}
```
This should target the same OCaml versions as dune `>= "4.08"`

tmcgilchrist commented on deps.ml:
```diff
@@ -0,0 +1,185 @@
+  (* After the d_items are filtered to the corresponding package request,
+   * we need to include the internal_deps in order to reach all the deps.
+   * If the internal dep is a public library we skip the recursive resolve
+   * because it will be resolve with separate request*)
```
```suggestion
   * because it will be resolve with separate request *)
```

tmcgilchrist commented on deps.ml:
```diff
@@ -0,0 +1,185 @@
+      d_items
+      |> List.filter (fun d_item ->
+          let item = Describe_external_lib.get_item d_item in
+          (* if an item has not package, we assume it's used for testing*)
```
```suggestion
          (* if an item has no package, we assume it's used for testing *)
```

tmcgilchrist commented on dune_items.ml:
```diff
@@ -0,0 +1,201 @@
+
+module Describe_opam_files = struct
+
+  type t = (string * OpamFile.OPAM.t ) list
+
+  let decode_items = function
+    | Sexp.List sexps ->
+      sexps
+      |> List.map (function
+          | Sexp.List [Atom opam_file; Atom opam_content] ->
+            (opam_file, OpamFile.OPAM.read_from_string opam_content)
+          | s -> Fmt.failwith "%s is not a good format decoding an item" (Sexp.to_string s))
+    | s -> Fmt.failwith "%s is not a good format decoding items" (Sexp.to_string s)
+
+  let opam_files_of_sexp = decode_items
```
```suggestion
  (** String representing an opam_file name eg. foo.opam *)
  type opam_file = string
  
  type t = (opam_file * OpamFile.OPAM.t ) list

   let opam_files_of_sexp = function
    | Sexp.List sexps ->
      sexps
      |> List.map (function
          | Sexp.List [Atom opam_file; Atom opam_content] ->
            (opam_file, OpamFile.OPAM.read_from_string opam_content)
          | s -> Fmt.failwith "%s is not a good format decoding an item" (Sexp.to_string s))
    | s -> Fmt.failwith "%s is not a good format decoding items" (Sexp.to_string s)
```

What about this?
Add a comment on the types (especially the opam_file string) and the entry function describing what it is for. Is the string the filename or a path?

tmcgilchrist replied:
What about this with comments on the types and an explanation of the opam_files_of_sexp function?
```suggestion
   (** String representing an opam_file name eg. foo.opam *)
  type opam_file = string

  (** ??? *)
  type t = (opam_file * OpamFile.OPAM.t ) list

  (** Decode opam_files from dune describe output. *)
  let opam_files_of_sexp = function
    | Sexp.List sexps ->
      sexps
      |> List.map (function
          | Sexp.List [Atom opam_file; Atom opam_content] ->
            (opam_file, OpamFile.OPAM.read_from_string opam_content)
          | s -> Fmt.failwith "%s is not a good format decoding an item" (Sexp.to_string s))
    | s -> Fmt.failwith "%s is not a good format decoding items" (Sexp.to_string s)
```

tmcgilchrist commented on dune_items.ml:
```diff
@@ -0,0 +1,201 @@
+      |> List.map (fun (package, entries) ->
+          (package, List.map (function
+               | Bin item   -> Bin {item with package = package}
+               | Other item -> Other {item with package = package}) entries))
```
Map over the sexp list this once rather than twice.
```suggestion
      sexps
      |> List.map (fun x-> decode_entries x |> (fun (package, entries) ->
          (package, List.map (function
               | Bin item   -> Bin {item with package = package}
               | Other item -> Other {item with package = package}) entries)))
```

tmcgilchrist commented on dune_items.ml:
```diff
@@ -0,0 +1,201 @@
+    Astring.String.cut ~sep:"/" ~rev:true s
+    |> Option.map snd |> Option.get
+
+  (* With "defautl/lib/bin.exe", it gives "default/lib" *)
```
```suggestion
  (* With "default/lib/bin.exe", it gives "default/lib" *)
```

tmcgilchrist replied:
```suggestion
  (* With "default/lib/bin.exe", it gives "default/lib" *)
```

tmcgilchrist commented on dune_items.ml:
```diff
@@ -0,0 +1,201 @@
+  module Kind = struct
+    type t = Required | Optional
+
+    let merge x y =
+      match (x, y) with
+      | Required,_ | _, Required -> Required
+      | Optional,Optional -> Optional
```
```suggestion
    let merge = function
      | Required,_ | _, Required -> Required
      | Optional,Optional -> Optional
```

tmcgilchrist replied:
```suggestion
    let merge x y = function 
      | Required,_ | _, Required -> Required
      | Optional,Optional -> Optional
```

tmcgilchrist commented on dune_items.ml:
```diff
@@ -0,0 +1,201 @@
+open Types
+
+module  Describe_external_lib = struct
```
```suggestion
module Describe_external_lib = struct
```

tmcgilchrist commented on dune_items.ml:
```diff
@@ -0,0 +1,201 @@
+  let is_lib_item = function
+    | Lib _ -> true | _ -> false
+
+  let string_of_atom =
+    function
+    | Sexp.Atom s -> s
+    | s -> Fmt.failwith "%s is an atom" (Sexp.to_string s)
```
```suggestion
  let string_of_atom = function
    | Sexp.Atom s -> s
    | s -> Fmt.failwith "%s is an atom" (Sexp.to_string s)
```

tmcgilchrist replied:
```suggestion
  let string_of_atom = function
    | Sexp.Atom s -> s
    | s -> Fmt.failwith "%s is an atom" (Sexp.to_string s)
```

tmcgilchrist commented on dune_items.ml:
```diff
@@ -0,0 +1,201 @@
+  let string_of_atom =
+    function
+    | Sexp.Atom s -> s
+    | s -> Fmt.failwith "%s is an atom" (Sexp.to_string s)
```
```suggestion
  let string_of_atom = function
    | Sexp.Atom s -> s
    | s -> Fmt.failwith "%s is an atom" (Sexp.to_string s)
```

tmcgilchrist commented on README.md:
```diff
@@ -26,18 +26,6 @@ Write changes? [y] y
-   opam package added its `META` file.
-6. Checks that each required opam package is listed in the opam file.
-7. For any missing packages, it offers to add a suitable dependency, using the installed package's version as the default lower-bound.
-
```
If this is no longer the case, can it be replaced with an explanation of how it is working now?

tmcgilchrist commented on dune-project:
```diff
@@ -16,9 +16,8 @@
-  (dune (< 3.0))
-  (ocaml (>= 4.10.0))
+  (stdune (>= 3.10.0))
+  (ocaml (>= 4.13.0))
```
This should match the dune ocaml version of 4.08

tmcgilchrist reviewed (commented):
Keeping compatibility with the same OCaml version as Dune is important. Please fix that up before merging.

moyodiallo committed:
Set the minimal version of OCaml to 4.08.0

moyodiallo committed:
Refactoring the code

moyodiallo committed:
Fix the update of opam files

moyodiallo committed:
check if the dune constraint matches the dune-project file

moyodiallo committed:
Add some tests

moyodiallo commented:
This issue https://github.com/ocurrent/opam-dune-lint/issues/28 is now fixed in the current PR.

cc @tmcgilchrist

MisterDA commented on dune_items.ml:
```diff
@@ -0,0 +1,206 @@
+    | s -> Fmt.failwith "%s is an atom" (Sexp.to_string s)
+
+  (* With "default/lib/bin.exe" or "default/lib/bin.bc.js" gives bin, it gives "bin.exe" *)
+  let bin_name s =
+    Astring.String.cut ~sep:"/" ~rev:true s
+    |> Option.map snd |> Option.get
```
I think you want [`Filename.basename`](https://v2.ocaml.org/api/Filename.html#VALbasename). Using `/` as a dir sep won't work on Windows, where it's `\`.

MisterDA commented on dune_items.ml:
```diff
@@ -0,0 +1,206 @@
+    |> Option.map snd |> Option.get
+
+  (* With "default/lib/bin.exe", it gives "default/lib" *)
+  let source_dir s =
+    Astring.String.cut ~sep:"/" ~rev:true s
+    |> Option.map fst |> Option.map (Astring.String.cut ~sep:"/" ~rev:false) |> Option.join |> Option.map snd
+    |> function None -> "." | Some dir -> dir
```
I think you want [`Filename.dirname`](https://v2.ocaml.org/api/Filename.html#VALdirname).

MisterDA commented on dune_items.ml:
```diff
@@ -0,0 +1,206 @@
+          List.map (fun name -> {dump_item with name = name}) (List.map string_of_atom sexps)
+        | _ -> items) [] sexps
+    in
+    List.fold_left (fun items sexps ->
+        match sexps with
```
```suggestion
    List.fold_left (fun items -> function
```

MisterDA commented on dune_items.ml:
```diff
@@ -0,0 +1,206 @@
+
+  let decode_item sexps =
+    let items =
+      List.fold_left (fun items sexps ->
+        match sexps with
```
```suggestion
      List.fold_left (fun items -> function
```

MisterDA commented on dune_items.ml:
```diff
@@ -0,0 +1,206 @@
+
+  let decode_items : Sexp.t list -> entry list =
+    List.filter_map (function
+        | Sexp.List [List [Atom "source"; List [Atom "User" ; _ ]]; List [Atom "entry"; List sexps]] -> Some (decode_item sexps)
+        | Sexp.List [List [Atom "entry"; List sexps]; List [Atom "source"; Atom "User"]] -> Some (decode_item sexps)
```
```suggestion
        | Sexp.List [List [Atom "source"; List [Atom "User" ; _ ]]; List [Atom "entry"; List sexps]]
        | Sexp.List [List [Atom "entry"; List sexps]; List [Atom "source"; Atom "User"]] -> Some (decode_item sexps)
```

MisterDA commented on dune_items.ml:
```diff
@@ -0,0 +1,206 @@
+  let entries_of_sexp : Sexp.t -> t list = function
+    | Sexp.List sexps ->
+      sexps
+      |> List.map (fun x -> decode_entries x |> (fun (package, entries) ->
+          (package, List.map (function
```
```suggestion
      |> List.map (fun x ->
          let package, entries = decode_entries x in
          (package, List.map (function
```

MisterDA commented on dune_items.ml:
```diff
@@ -0,0 +1,206 @@
+
+  let items_bin_of_entries describe_entries =
+    List.concat_map snd describe_entries
+    |> List.filter_map (fun d_item ->
+        d_item
+        |> (function Bin item -> Some (item.bin_name,item) | Other _ -> None))
```
```suggestion
    |> List.filter_map (function Bin item -> Some (item.bin_name,item) | Other _ -> None)
```

MisterDA commented on main.ml:
```diff
@@ -156,19 +194,30 @@ let main force dir =
+  get_opam_files ()
+  |> Paths.to_seq
+  |> List.of_seq
+  |> List.map (fun (path, opam) ->
+      let pkg_name = (OpamPackage.Name.of_string (Filename.chop_suffix path ".opam")) in
+      Dune_constraints.check_dune_constraints ~errors:[] ~dune_version pkg_name opam)
+  |> List.flatten
```
```suggestion
  |> List.concat_map (fun (path, opam) ->
      let pkg_name = (OpamPackage.Name.of_string (Filename.chop_suffix path ".opam")) in
      Dune_constraints.check_dune_constraints ~errors:[] ~dune_version pkg_name opam)
```

moyodiallo replied:
I have to avoid `concat_map` because introduced sine OCaml.4.10. We need to be compatible with ocaml.4.08 because of the minimal set of dune.

moyodiallo replied:
I get it. 😄

MisterDA commented on deps.ml:
```diff
@@ -0,0 +1,185 @@
+  | Some x -> x
+  | None ->
+    let r =
+      match Astring.String.cut ~sep:"/" ~rev:true path with
```
Won't work on Windows, the file separator isn't `/`. Consider using [`Fpath`](https://erratique.ch/software/fpath/doc/Fpath/index.html).

MisterDA commented on deps.ml:
```diff
@@ -0,0 +1,185 @@
+  let d_items_lib =
+    d_items
+    |> List.filter_map (fun d_item ->
+        match is_lib_item d_item with
+        | true ->
+          d_item
+          |> get_item
+          |> (fun (item:Describe_external_lib.item) ->
+              (String.cat item.name ".lib", Lib item))
+          |> Option.some
+        | false -> None)
```
Prefer using `if...then...else` rather than `match` expressions on booleans. The rest of the code can be made clearer like this, imo.

```suggestion
        if is_lib_item d_item then
          let item = get_item d_item in
          Some (item.Describe_external_lib.name ^ ".lib", Lib item)
        else None)
```

MisterDA commented on deps.ml:
```diff
@@ -0,0 +1,185 @@
+      let item = Describe_external_lib.get_item d_item in
+      if Describe_external_lib.is_exe_item d_item && Option.is_none item.package
+      then
+        match find_package_of_exe item  with
+        | None ->  d_item
```
```suggestion
        match find_package_of_exe item with
        | None -> d_item
```

MisterDA commented on deps.ml:
```diff
@@ -0,0 +1,185 @@
+      in d_items, unresolved_entries)
+  |> (fun (d_items, unresolved_entries) ->
+      d_items
+      |> List.map (fun d_item ->
+          match d_item with
+          | Describe_external_lib.Exe item ->
```
```suggestion
      |> List.map (function
          | Describe_external_lib.Exe item as d_item ->
```

MisterDA reviewed (commented):
Some suggestions on coding style.

moyodiallo committed:
Refactoring the source code

MisterDA commented on CHANGES.md:
```diff
@@ -1,3 +1,8 @@
+### Unreleased
+
+- Add support for dune 3.0 , the command `dune external-lib-deps` was remove from
+  dune. Now, `opam-dune-lint` command works without `dune build`. (@moyodiallo #46).
```
```suggestion
- Add support for dune 3.0 , the command `dune external-lib-deps` was removed from
  dune. Now, the `opam-dune-lint` command works without `dune build`. (@moyodiallo #46).
```

MisterDA commented on README.md:
```diff
@@ -29,14 +29,15 @@ Wrote "dune-project"
-4. Filters out vendored dependencies (by ignoring dependencies from subdirectories with their own `dune-project` file).
-5. For each ocamlfind library, it finds the corresponding opam library by
+2. Runs `dune describe external-lib-deps` to get all externals and internals ocamlfind libraries for all dune libraries, executables and tests. The information about the package is also known except for the private executables.
+3. Runs `dune describe package-entries` to get all packages entries, this is for considering the external ocamlfind libraries of a private executable, because in dune it possible to install an private executable.
```
```suggestion
3. Runs `dune describe package-entries` to get all packages entries, this is for considering the external ocamlfind libraries of a private executable, because in Dune it is possible to install an private executable.
```

MisterDA commented on deps.ml:
```diff
@@ -0,0 +1,185 @@
+       d_item
+       |> Describe_external_lib.get_item
+       |> (fun (item:Describe_external_lib.item) -> Fpath.(item.source_dir / "dune"))
+       |> (Dune_rules.Copy_rules.get_copy_rules))
```
```suggestion
       |> Dune_rules.Copy_rules.get_copy_rules)
```

MisterDA commented on deps.ml:
```diff
@@ -0,0 +1,185 @@
+  get_dune_items (Hashtbl.create 10) ~pkg ~target
+  |> List.map Describe_external_lib.get_item
+  |> List.fold_left (fun libs (item:Describe_external_lib.item) ->
+      List.map (fun dep -> fst dep, item.source_dir) item.external_deps
```
This removes one map over the list, right?
```suggestion
  |> List.fold_left (fun libs item ->
      let item = Describe_external_lib.get_item item in
      List.map (fun dep -> fst dep, item.source_dir) item.external_deps
```

MisterDA commented on dune_items.ml:
```diff
@@ -0,0 +1,197 @@
+      package = None;
+      external_deps = [];
+      internal_deps = [];
+      source_dir = Fpath. v ".";
```
```suggestion
      source_dir = Fpath.v ".";
```

MisterDA commented on dune_items.ml:
```diff
@@ -0,0 +1,197 @@
+
+  let string_of_atom = function
+    | Sexp.Atom s -> s
+    | s -> Fmt.failwith "%s is an atom" (Sexp.to_string s)
```
```suggestion
    | s -> Fmt.failwith "%a is an atom" Sexp.pp_hum s
```
This might give better human-readable output. You may want to change error-logging to use these `pp_hum` (pretty-print for humans) functions.
https://ocaml.org/p/sexplib/latest/doc/Sexplib/Sexp/index.html#output-of-s-expressions-to-formatters

MisterDA commented on deps.ml:
```diff
@@ -0,0 +1,185 @@
+  if Fpath.is_current_dir path then false
+  else
+    Fpath.(path / "dune-project")
+    |>  Bos.OS.Path.exists
```
```suggestion
    |> Bos.OS.Path.exists
```

MisterDA commented on dune_project.ml:
```diff
@@ -98,149 +94,42 @@ let update (changes:(_ * Change.t list) Paths.t) (t:t) =
+     | Some version -> version
+
+let dune_format dune =
+  Bos.OS.Cmd.(in_string dune |>  run_io Bos.Cmd.(v "dune" % "format-dune-file") |> out_string)
```
```suggestion
  Bos.OS.Cmd.(in_string dune |> run_io Bos.Cmd.(v "dune" % "format-dune-file") |> out_string)
```

MisterDA commented on dune_rules.ml:
```diff
@@ -0,0 +1,88 @@
+    in
+    sexps
+    |> List.filter is_action_copy
+    |> List.map (fun rule ->
```
`List.filter_map` here?

MisterDA commented on dune_rules.ml:
```diff
@@ -0,0 +1,88 @@
+
+  let get_copy_rules file =
+    match Hashtbl.find_opt rules file with
+    | None when  OS.Path.exists file |> Result.get_ok ->
```
```suggestion
    | None when OS.Path.exists file |> Result.get_ok ->
```

MisterDA commented on main.ml:
```diff
@@ -135,18 +143,48 @@ let confirm_with_user () =
+      else
+        (* prevent `dune describe opam-files` crashing when there is a opam file `*.opam`
+         * and its package description is missing in dune-project file.*)
+        Sys.remove path) old_opam_files;
```
```suggestion
      (* prevent `dune describe opam-files` crashing when there is a opam file `*.opam`
       * and its package description is missing in dune-project file.*)
      if not (Paths.mem path packages) then
        Sys.remove path) old_opam_files;
```

MisterDA commented on types.ml:
```diff
@@ -12,6 +22,21 @@ module Change = struct
 
+module List = struct
+    include List
+    let concat_map f l = List.map f l |> List.flatten
```
That's not how you should implement `concat_map`, this is not tail-recursive and removes the benefit of using `List.concat_map`.

moyodiallo replied:
The tail-recursive is not a problem here, we are not manipulating huge list.

MisterDA commented on types.ml:
```diff
@@ -29,7 +54,9 @@ module Change_with_hint = struct
   let pp f (c, dirs) =
-    let dirs = Dir_set.map (function "." -> "/" | x -> x) dirs in
+    let dirs =
+      Dir_set.map (fun path -> if Fpath.is_current_dir path then Fpath.v "/" else path) dirs
```
Are you trying to get the root directory? This won't work on Windows, where the root is not `/` (nor `\` for that matter. If you _really_ need the root, you could get the current working directory with `Sys.getcwd ()` and extract the root from that path. If you have another way to do what you want without using the root, you should try that. If you want the directory separator you could use `Filename.dir_sep` or `Fpath.dir_sep` for clarity.

moyodiallo replied:
The purpose is to print information. This is not supposed to be used.

MisterDA commented on types.ml:
```diff
@@ -48,3 +75,17 @@ module Change_with_hint = struct
+  |> (fun s ->
+      try Sexp.of_string s with
+      | Sexp.Parse_error _ as e ->
+        Fmt.pr "Error parsing '%s' output:\n" (Bos.Cmd.to_string cmd); raise e)
```
Maybe it'd be better to log on standard error?

moyodiallo replied:
Would you like to give an example ?

MisterDA replied:
I was thinking of [`Fmt.epr`](https://erratique.ch/software/fmt/doc/Fmt/index.html#val-epr) but that's really up to you :)

moyodiallo replied:
I get it.

moyodiallo committed:
Refactoring the source code

moyodiallo committed:
Merge branch 'master' into dune-external-libs

smorimoto commented:
When will this be released?

moyodiallo commented:
We are testing it on https://ocaml.ci.dev/ via `opam lint` jobs. When it seems stable, it will be released.
